### PR TITLE
Build: sanitize init system + launchers handling (incl. spec), and more

### PR DESCRIPTION
**Pull request by jnpkrn**

### CODE REVIEW

**Comment by jnpkrn** [comment]
Fixed on more issue, this time with `s/upstart/crm_mon`, and also realized there's a trap in that one cannot use snippets like this:

```
%if %{defined _unitdir}
BuildRequires: systemd-devel
%endif
```

as it introduces chicken-and-egg problem (`_unitdir` macro cannot be queried reliably at before-build-requires-installed time as we only know that macro exists once `systemd` package was build-required).

**Comment by kgaillot** [comment]
I don't think init launcher support should be exclusive. For example, Gentoo allows users to select LSB or systemd. It would make sense to make a Gentoo build that deploys both the unit file and the init script, and lets the user use their installed init launcher to enable the right one (or even switch launchers in the future, without having to rebuild pacemaker). I can also imagine cases where someone uses systemd as their default launcher, but prefers to start certain services via some other mechanism that requires the init script. These may be outliers, but I don't think we should actively prevent them.

**Comment by jnpkrn** [comment]
On 13/10/16 13:13 -0700, Ken Gaillot wrote:
> I don't think init launcher support should be exclusive. For
> example, Gentoo allows users to select LSB or systemd. It would make
> sense to make a Gentoo build that deploys both the unit file and the
> init script, and lets the user use their installed init launcher to
> enable the right one (or even switch launchers in the future,
> without having to rebuild pacemaker). I can also imagine cases where
> someone uses systemd as their default launcher, but prefers to start
> certain services via some other mechanism that requires the init
> script. These may be outliers, but I don't think we should actively
> prevent them.

Don't consider this a good argument:

- Gentoo is flag-driven, so the ebuild can/should conditionalize based
  on the respective flags whether to use unit files or init scripts,
  plus this is really a special group (I used to be one of them)

- other use cases are moot as well, it's easy to ask the build system
  to prepare whatever they want, if not picking it from the tarball
  (git) directly

Let's solve this on when-it-comes real basis, rather than continuing
in this wannabe versatility madness while not doing the things
properly arising from the self-inflicted complexity (see
"%bcond_with upstart_job" in the spec).

-- 
Jan (Poki)

**Comment by jnpkrn** [comment]
Regarding the mentioned trap, I've opened
https://github.com/rpm-software-management/mock/issues/11
to see what's possible and what's not in this special systemd case.

**Comment by jnpkrn** [comment]
It was confirmed that this is indeed by design.  I wasn't aware of this till now, and it also invalidates some assumptions in the spec so far as mentioned.

Anyway, I seem to have overcome that, so in turn, `pkgconfig(dbus-1)` as a BR is installed only when suitable, which was discussed in the past.

**Comment by jnpkrn** [comment]
Test Copr builds: https://copr.fedorainfracloud.org/coprs/jpokorny/pacemaker-testing/build/465155/

File: pacemaker.spec.in [hunk 1]
```diff
@@ -28,34 +28,35 @@
...
 ## Where to install python site libraries (currently, this uses the unversioned
 ## python_sitearch macro to get the default system python, but at some point,
 ## we should explicitly choose python2_sitearch or python3_sitearch -- or both)
-%global py_site    %{?python_sitearch}%{!?python_sitearch:%(python -c "from distutils.sysconfig import get_python_lib; print(get_python_lib(1))")}
+%define py_site %{?python_sitearch}%{!?python_sitearch:%(
+  python -c 'from distutils.sysconfig import get_python_lib as gpl; print(gpl(1))' 2>/dev/null)}
 
 ## Whether this platform defaults to using CMAN
-%global cman_native 0%{?el6} || (0%{?fedora} > 0 && 0%{?fedora} < 17)
+%define cman_native (0%{?el6} || (0%{?fedora} > 0 && 0%{?fedora} < 17))
```
Comment (kgaillot) [inline pacemaker.spec.in]: Why make cman_native and debug_package defines?

File: pacemaker.spec.in [hunk 2]
```diff
@@ -176,6 +176,9 @@ BuildRequires: python-devel >= 2.6
 Requires:      libqb >= 0.13.0
 BuildRequires: libqb-devel >= 0.13.0
 
+# Basics required for the build (even if usually satisfied through other BRs)
+BuildRequires: coreutils findutils grep sed
```
Comment (kgaillot) [inline pacemaker.spec.in]: BTW when adding new dependencies, we should check whether the package names are the same on suse. The GNUmakefile target for *-suse.spec actually uses sed to replace package names (a questionable approach, but one we want to stick with for the short term).

File: configure.ac [hunk 3]
```diff
@@ -372,23 +372,6 @@ case $CSPREFIX in
   prefix) CSPREFIX=$prefix;;
 esac
 
-AC_MSG_NOTICE(Sanitizing INITDIR: ${INITDIR})
-case $INITDIR in
-  prefix) INITDIR=$prefix;;
-  "")
-    AC_MSG_CHECKING(which init (rc) directory to use)
-      for initdir in /etc/init.d /etc/rc.d/init.d /sbin/init.d	\
-	   /usr/local/etc/rc.d /etc/rc.d
```
Comment (kgaillot) [inline configure.ac]: This isn't "overly smart" -- it's a simple list of common locations, and avoids unnecessary extra work for our users to figure out local details, which is the point of the configure script. Someone building pacemaker (possibly in a heterogeneous environment) shoudn't have to be aware of this option or find an appropriate value, for the common case of running "./configure".

Comment (jnpkrn) [inline configure.ac]: One of the point of configure script is to provide interface alleviating the build system from second-guessing of how to result should look like with regards to the target system.  Doing so based on basically random existence of particular directories is a perfect example of "overly smart second-guessing".  When building in a minimalistic chroot (such as with RPM ecosystem using mock heavily), they may not even exist at all.  So if very basic `/etc/init.d` wvon't fit, (fairly advanced by definition) user is supposed to specify the location via `--with-initdir=DIR`.

What's the conflict?  What the surprises are there when help screen navigates one to do so?

If you cannot follow these intuitive steps, you should delegate the building question to someone else, really.

**Comment by jnpkrn** [comment]
On 18/10/16 12:56 -0700, Ken Gaillot wrote:
> Gentoo is an example, not a reason. The point is that users should
> be able to build a single pacemaker package that can be deployed in
> a heterogeneous environment utilizing various launchers, as well as
> remain usable if the system it's on has its launch method changed.

So, what are we discussing now, once again -- result of "make install"
or result of building an RPM package using the upstream spec file?

If the former, it's not expected that you'll be disappointed by
missing initscript when you have systemd around.  If you are that
special, you likely know what you are doing and it's trivial to
copy respective files from the git/tarball on your own.

If the latter, you already have to deal with burden that naturally
arise from heterogenous environment:

- different architectures
- different target distros with varying software present, so you
  will be bound to particular library versions (RPM-tracked
  SONAME requirements), etc.
 
so I don't think being bound to particular init system makes it any
worse (to the point this is just a philosophical debate).  You want
to have as little distributions flying in your "heterogenous"
environment as possible, especially for HA where identical pieces
of software and hardware makes root cause analysis of problems
easier, which also minimizes number of packages of the same component
you really need.

> We are also hoping to make pacemaker able to be deployed in a
> container, which may launch in an unusual way -- having the LSB
> script available may be important, even if the underlying OS is
> systemd native.

I suppose RPM itself will be of less importance for container
deployments (will it rather mean pre-cooked images?) and I have no
ideas how the new recipes will look like, but again, copying files
around should be non-issue.

> More generally, this is already a real capability we provide to
> users, and I don't see any reason to take it away from them.

Or seen from the other side, it's a burden installed even if it makes
little sense outside of initscript-driven system.  Nobody takes anyone
the liberty to grab those files, but I don't suppose there would be
justified demand for that to always be present.

> Also, I'm not sure there's an advantage in using initdir in the
> context of systemd and upstart. I don't think users would
> intuitively recognize the usage. Having separate means of specifying
> systemd/upstart directories, and being able to build a package that
> supports any desired combination of launch methods, makes more
> sense.

That's the exploding complexity I was talking about for little to no
material gain.  Help screen message for that item was modified
respectively, and the logic behind is also readable from the output
of configure script itself.

-- 
Jan (Poki)

**Comment by kgaillot** [comment]
I've been slammed lately and haven't been able to consider this ... can you submit 997d58e, 54eed6e, 214d932, fea9afd, and cc839b9 as a separate PR so I can merge the rest here?

Re: 18d124a: unless I'm misunderstanding, I believe cman_native and debug_package can remain globals, to avoid unnecessary re-evaluation.

**Comment by jnpkrn** [comment]
Closed in favor of #1175, some commits already included in #1165.

**Pull request closed without merging**